controllers/mr_inferenceservice_controller.go

```diff
@@ -0,0 +1,414 @@
+	is := &openapi.InferenceService{
+		Name:                 &isName,
+		ServingEnvironmentId: servingEnvironmentId,
+		RegisteredModelId:    registeredModelId,
+		ModelVersionId:       modelVersionIdPtr,
+		Runtime:              isvc.Spec.Predictor.Model.Runtime,
+		DesiredState:         openapi.INFERENCESERVICESTATE_DEPLOYED.Ptr(),
+	}
+
+	return mr.UpsertInferenceService(is)
+}
+
+// onDeletion mark model registry inference service to UNDEPLOYED desired state
+func (r *ModelRegistryInferenceServiceReconciler) onDeletion(mr api.ModelRegistryApi, log logr.Logger, isvc *kservev1beta1.InferenceService, is *openapi.InferenceService) error {
+	log.Info("Running onDeletion logic")
+	is.DesiredState = openapi.INFERENCESERVICESTATE_UNDEPLOYED.Ptr()
+
+	_, err := mr.UpsertInferenceService(is)
+	return err
+}
+
+// initModelRegistryService setup a gRPC connection with MLMD server and initialize the model registry service
+func (r *ModelRegistryInferenceServiceReconciler) initModelRegistryService(ctx context.Context, ctxTimeout context.Context, log logr.Logger, namespace string) (api.ModelRegistryApi, *grpc.ClientConn, error) {
+	log1 := log.WithValues("mr-namespace", namespace)
+	mlmdAddr, ok := os.LookupEnv(constants.MLMDAddressEnv)
+
+	if !ok || mlmdAddr == "" {
+		log1.Info("Retrieving mlmd address from deployed model registry service")
+		// Env variable not set, look for existing model registry service
+		opts := []client.ListOption{client.InNamespace(namespace), client.MatchingLabels{
+			"component": "model-registry",
+		}}
+		mrServiceList := &corev1.ServiceList{}
+		err := r.client.List(ctx, mrServiceList, opts...)
+		if err != nil || len(mrServiceList.Items) == 0 {
+			// No model registry deployed in the provided namespace, skipping serving reconciliation
+			return nil, nil, fmt.Errorf("unable to find model registry in the given namespace: %s", namespace)
+		}
+
+		// Actually we could iterate over every mrService, as nothing prevents to setup multiple MR in the same namespace
+		if len(mrServiceList.Items) > 1 {
+			return nil, nil, fmt.Errorf("multiple services with component=model-registry for Namespace %s", namespace)
+		}
+
+		mrService := mrServiceList.Items[0]
+
+		var grpcPort *int32
+		for _, port := range mrService.Spec.Ports {
+			if port.Name == "grpc-api" {
+				grpcPort = &port.Port
+				break
+			}
+		}
+
+		if grpcPort == nil {
+			return nil, nil, fmt.Errorf("cannot find grpc-api port for service %s", mrService.Name)
+		}
+
+		mlmdAddr = fmt.Sprintf("%s.%s.svc.cluster.local:%d", mrService.Name, namespace, *grpcPort)
+	}
+
+	if mlmdAddr == "" {
+		log1.Info("Cannot connect to the model registry: empty mlmd address")
+		return nil, nil, nil
+	}
+
+	// Setup model registry service
+	log.Info("Connecting to " + mlmdAddr)
+	conn, err := grpc.DialContext(
+		ctxTimeout,
+		mlmdAddr,
+		grpc.WithReturnConnectionError(),
+		grpc.WithBlock(),
```

Comment:
Hey @israel-hdez , thanks for this pointer.. that's interesting, based on the link you provided it seems that the main reasons it is discouraged is that:
1. The connection check is done at dial time ([ref](https://github.com/grpc/grpc-go/blob/master/Documentation/anti-patterns.md#using-failonnontempdialerror-withblock-and-withreturnconnectionerror:~:text=The%20options%20FailOnNonTempDialError%2C%20WithBlock%2C%20and%20WithReturnConnectionError%20are%20designed%20to%20handle%20some%20of%20these%20errors%2C%20but%20they%20do%20so%20by%20relying%20on%20failures%20at%20dial%20time)) therefore there is no guarantee the connection won't fail in subsequent requests --> We are already managing this as every time we perform a request we check if that returned an error or not (either logic error or connection one)
2. Wait for connection forever ([ref](https://github.com/grpc/grpc-go/blob/master/Documentation/anti-patterns.md#using-failonnontempdialerror-withblock-and-withreturnconnectionerror:~:text=if%20a%20client%20uses%20WithBlock%20to%20wait%20for%20a%20connection%20to%20be%20established%2C%20it%20may%20end%20up%20waiting%20indefinitely%20if%20the%20server%20is%20not%20responding)) --> we are already managing this case by using a context timeout to avoid infinite wait time.

wdyt?

Reply:
So, in the updated code, I see you removed both `WithReturnConnectionError` and `WithBlock`, and I no longer see ctxTimeout. To me is better now, as we follow the recommendations.

Just to answer your thoughts:
1. Yes, if you are handling errors, I see OK to keep `WithReturnConnectionError`.
2. The (2) was my main concern, as I don't know if `WithBlock` honors the timeout, or if it blocks indefinitely regardless of the configured timeout in the Context. The [api docs](https://pkg.go.dev/google.golang.org/grpc#WithBlock) don't clarify it, so I am unsure of the actual behavior.